---
title: "ThunderBio scRNA-seq Report"
date: "`r Sys.Date()`"
output: 
  flexdashboard::flex_dashboard:
    vertical_layout: scroll
    orientation: rows
    theme: 
      bootswatch: cosmo
      version: 5
      base_font:
        google: Asap
params:
  sampleName : "sampleName"
  featureStats : "sampleID.featureCoverage_stats.json"
  geneCoverage : "sampleID.scaled.tab"
  starsolo_matrixDir: "sample/Gene/filtered"
  starsolo_out : "Summary.csv"
  starsolo_bc : "UMIperCellSorted.txt"
  saturation_json: "saturation_out.json"
  version_json: "versions.json"
  nCPUs: 10
  nMem: 10000000000.0
  nFeature_RNA_lower: 200
  percent_mt_cutoff: 100
  pca_dim: 30
  resolution: 0.8
---

```{r setup, include=FALSE}
library(tidyverse)
library(scales)
library(flexdashboard)
library(jsonlite)
library(DT)
library(htmltools)
##library(htmlwidgets)
library(plotly)
library(kableExtra)
library(Seurat)
library(future)

set.seed(1234)
plan(multisession, workers = min(as.numeric(params$nCPUs),4))
options(future.globals.maxSize = as.numeric(params$nMem), sass.cache = FALSE)
RhpcBLASctl::blas_set_num_threads(1) # https://github.com/satijalab/seurat/issues/3991

## Rmd options
knitr::opts_chunk$set(
  comment = '', fig.retina = 4,
  warning = FALSE, message = FALSE
)

html_font <- "Asap, Arial, sans-serif"
```

```{r infoIcon, include=FALSE}
infoIcon <- function(targetEl, size = "1.2rem"){
  querySelector <- paste0("#", targetEl)
  ## Make HTML
  questionIconString <- paste0(
      '<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 16 16" class="bi bi-question-circle-fill " style="height:',
      size,
      ';width:',
      size,
      ';fill:currentColor;vertical-align:-0.125em;fill:var(--bs-primary);" aria-hidden="true" role="img"><path d="M16 8A8 8 0 1 1 0 8a8 8 0 0 1 16 0zM5.496 6.033h.825c.138 0 .248-.113.266-.25.09-.656.54-1.134 1.342-1.134.686 0 1.314.343 1.314 1.168 0 .635-.374.927-.965 1.371-.673.489-1.206 1.06-1.168 1.987l.003.217a.25.25 0 0 0 .25.246h.811a.25.25 0 0 0 .25-.25v-.105c0-.718.273-.927 1.01-1.486.609-.463 1.244-.977 1.244-2.056 0-1.511-1.276-2.241-2.673-2.241-1.267 0-2.655.59-2.75 2.286a.237.237 0 0 0 .241.247zm2.325 6.443c.61 0 1.029-.394 1.029-.927 0-.552-.42-.94-1.029-.94-.584 0-1.009.388-1.009.94 0 .533.425.927 1.01.927z"></path></svg>'
  )
  outTag <- tags$i(
      HTML(questionIconString),
      `data-bs-toggle`="modal",
      `data-bs-target`=querySelector
  )
  return(outTag)
}

createModal <- function(Id, expl_tbl){
    contentTag <- expl_tbl %>%
        apply(1, FUN=function(x){
            tagList(
                h6(HTML(x[1]), style="font-weight:700;"),
                p(HTML(x[2]))
            )
        }) %>% 
        tagList()

   outTag <- div(
       class = "modal fade",
       id = Id,
       tabindex = -1,
       `aria-labelledby`="exampleModalLabel",
       `aria-hidden`="true",
       div(
           class = "modal-dialog modal-dialog-centered modal-dialog-scrollable modal-lg",
           div(
               class = "modal-content",
               div(
                   class = "modal-body",
                   div(
                       class = "explanation",
                       contentTag
                   )
               )
           )
       )
   )
   return(outTag)
}
```

```{css}
.explanation {
  background-color: rgba(var(--bs-secondary-rgb), 0.2); 
  border-radius: var(--bs-modal-border-radius);
  padding: 1rem
}
```

```{r modal, echo = FALSE}
expl_tbl <- tribble(
    ~title, ~content,
    "Estimated Number of Cells",
    "The number of barcodes detected as cells",
    "Unique Reads in Cells Mapped to Gene",
    "The number of reads uniquely mapped to gene region in cell associated barcodes",
    "Fraction of Unique Reads in Cells",
    'The fraction of reads uniquely mapped to gene region in cell associated barcodes (divided by total reads uniquely mapped to gene region in all of the barcodes). This value could be interpreted as the "signal to noise ratio"',
    "Mean Reads per Cell",
    "The mean number of reads in the cell associated barcodes",
    "Median Reads per Cell",
    "The median number of reads in the cell associated barcodes",
    "UMIs in Cells",
    "The total number of UMI counts in the cell associated barcodes",
    "Mean UMI per Cell",
    "The mean number of UMI counts in the cell associated barcodes",
    "Median UMI per Cell",
    "The median number of UMI counts in the cell associated barcodes",
    "Mean Gene per Cell",
    "The mean number of genes detected in the cell associated barcodes",
    "Median Gene per Cell",
    "The median number of genes detected in the cell associated barcodes",
    "Total Gene Detected",
    "Total number of genes detected in all of the cell associated barcodes"
)
createModal("cellStats", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Knee Plot",
    'The Knee Plot is also called Barcode Rank Plot. The graph displays the number of filtered UMIs assigned to each barcode. The blue color indicates cell associated barcodes, while grey ones are background barcodes. Cell associated barcodes are not solely determined by their UMI count, by default the pipeline will invoke the "EmpytDrops" method, which will compare the expression profile of pre-defined cells and the background, then rescue some "low UMI" cells.'
)
createModal("kneePlot", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Number of Reads",
    "Number of read pairs after trimming low quality nucleotides from the end of the cDNA read and discarding too short reads",
    "Reads With Valid Barcodes",
    "Percentage of reads with valid barcode sequence: <li>No mare than 2 N in barcode sequence</li><li>No homopolymer</li><li>Only one mismatch with whitelist</li><li>Reads matches a whitelist barcode with 1 mismatch, but this whitelist barcode does not get any other reads with exact matches of CB were discarded</li>",
    "Sequencing Saturation",
    "The saturation was calculated as 1-nUMI/nReads, which could be used to measure how many reads it will take to sequence a new transcript (UMI). nUMI is the number of UMI, nReads is the number of reads with valid barcode/UMI/Gene. This value represents the complexity of the library. For instance, if the saturation is 75%, then we could only get one new UMI when sequence four more reads.",
    "Q30 Bases in CB+UMI",
    "Percentage of the nucleotides in cell barcode and UMI sequence with the sequencing quality larger than 30 (99.9% accuracy)",
    "Q30 Bases in RNA read",
    "Percentage of the nucleotides in cDNA reads with sequencing quality larger than 30 (99.9% accuracy)"
)
createModal("sequencingStats", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Mapped to Genome: Unique+Multiple",
    "Total mapping rate of the input reads in spite of barcodes matching",
    "Reads Mapped to Genome: Unique",
    "Fraction of the reads uniquely mapped to the genome in spite of barcodes matching",
    "Reads Mapped to Gene: Unique+Multiple Gene",
    '<blockquote>Multi-gene reads are concordant with (i.e. align equally well to) transcripts of two or more genes. One class of multi-gene read are those that map uniquely to a genomic region where two or more genes overlap. Another class are those reads that map to multiple loci in the genome, with each locus annotated to a different gene.</blockquote> STARsolo will output "NoMulti" if multi-genes reads are not counted',
    "Reads Mapped to Gene: Unique Gene",
    'Fraction of the reads mapped to a unique gene region, this parameter will mostly equivalent to "Reads Mapped Confidently to Transcriptome" in cellranger report',
    "Reads Mapped to Exonic Regions",
    "Fraction of the reads mapped to exonic region",
    "Reads Mapped to Intronic Regions",
    "Fraction of the reads mapped to intronic region",
    "Reads Mapped to Intergenic Regions",
    "Fraction of the reads mapped to intergenic region"
)
createModal("mappingStats", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Cell Cluster (colored by cell group)",
    "The UMAP dimension reduction result of the cellular expression profile. Cells with similar expression were assigned to the same group using the unsupervised clustering algorithm of Seurat package. Each scatter represents a cells, colored by the cell group information."
)
createModal("clusterCellGroups", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Cell Cluster (colored by UMI density)",
    "The UMAP dimension reduction result of the cellular expression profile. Each scatter represents a cells, colored by the UMI counts detected (log-transformed)."
)
createModal("clusterUMIDensity", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Marker Genes of Clusters",
    "After dimension reduction, cells were grouped by unsupervised clustering method (leiden), group specifically expressed genes were tested and the top five genes were listed here."
)
createModal("DEGTable", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Cell Metrics",
    "The number of the genes, the UMIs, as well as the mitochondrial UMI percentage in each cell were summarized in a violin plot."
)
createModal("cellMetrics", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Gene Coverage",
    "The reads coverage accross the mRNA body. Typical 3' scRNAseq will have a clear peak at the transcript end, while 5' scRNAseq will have higher read counts at the mRNA start."
)
createModal("geneCoverage", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Sequencing Saturation",
    "The reads of the library were downsampled at different ratio to calculate the sequencing saturation at different sequencing depth. The saturation was calculated as 1-nUMI/nReads, which could be used to measure how many reads it will take to sequence a new transcript (UMI)"
)
createModal("sequencingDepthCurve", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Median Genes per Cell",
    "The reads of the library were downsampled at different ratio to simulate sequencing the library with lower depth, the median value of the genes detected in each cell was calculated to represent a gene-level saturation."
)
createModal("medianGeneCurve", expl_tbl)

expl_tbl <- tribble(
    ~title, ~content,
    "Sample ID",
    "ID of the sample",
    "Pipeline Version",
    "Version of the StarScope pipeline",
    "Reference Dir",
    "The reference genome STAR index file path",
    "Reference GTF",
    "Genome annotation file path",
    "STAR Version",
    "The version of the STAR program",
    "Whitelist Matching",
    "Mothod of matching the cell barcodes to the whiteList. By default we used <code>--soloCBmatchWLtype 1MM_multi_Nbase_pseudocounts</code> parameter, which allows multiple matches in whitelist with 1 mismatched base and a posterior probability will be calcuated to choose one of the matches. Pseudocounts of 1 are added to all whitelist barcodes, multimatching to WL is allowed for CBs with N-bases. This option matches best with CellRanger >= 3.0.0",
    "UMIfiltering",
    "UMI filtering method, by default <code>--soloUMIfiltering MultiGeneUMI_CR</code> will remove UMIs with N and homopolymers as well as lower-count UMIs that map to more than one gene, matching the calculation method in CellRanger > 3.0.0",
    "UMIdedup",
    "Type of UMI deduplication (collapsing) algorithm, by default CellRanger2-4 algorithm for 1MM UMI collapsing was employed, and <code>--soloUMIdedup 1MM_CR</code> parameter was used.",
    "Cell Calling",
    "Cell calling method used in the STARsolo run, by default we employed <b>EmptyDrop</b> method, and <code>--soloCellFilter EmptyDrops_CR</code> parameter was used.",
    "includeIntron",
    "Indicate if reads mapped to intron region were included in the analysis. If it is TRUE here, <code>--soloFeatures GeneFull</code> was used in the STARsolo run, otherwise <code>--soloFeatures Gene</code> was used.",
    "includeMultiReads",
    "Indicate if reads mapped to mulitple genes were included in the analysis, currently not supported and will always be FALSE."
)
createModal("runningInfo", expl_tbl)
```

```{r include=FALSE}
config_plotly_fig <- function(fig){
    config(
        fig,
        displaylogo = FALSE, 
        modeBarButtonsToRemove = c('zoom', 'pan', 'select', 'zoomIn', 'zoomOut', 'autoScale',
                                   'hoverClosestCartesian', 'hoverCompareCartesian'),
        toImageButtonOptions = list(height= NULL, width= NULL, scale= 2)
    )
}

is_number_string <- function(x){
    str_split(x, pattern="") %>% 
        unlist %>% str_detect("[0-9.]") %>% 
        all()
}
starsolo_summary <- read_csv(params$starsolo_out,
                             col_names = c("item", "value"),
                             col_types = cols("c", "c"))
starsolo_summary$data <- NA
for(i in 1:nrow(starsolo_summary)){
    if(is_number_string(starsolo_summary$value[i])){
        if(as.numeric(starsolo_summary$value[i]) >= 1){
            starsolo_summary$data[i] <- comma(as.numeric(starsolo_summary$value[i]), accuracy = 1)
        }else{
            starsolo_summary$data[i] <- label_percent(0.01)(as.numeric(starsolo_summary$value[i]))
        }
    }else{
        starsolo_summary$data[i] <- starsolo_summary$value[i]
    }
}
starsolo_summary <- starsolo_summary %>%
    select(-value) %>%
    dplyr::rename("value" = "data") %>%
    mutate(item=str_replace_all(item, "GeneFull", "Gene"))
    
exon_ratio <- fromJSON(params$featureStats)$exonRatio %>% as.numeric
intron_ratio <- fromJSON(params$featureStats)$intronRatio %>% as.numeric
intergenic_ratio <- fromJSON(params$featureStats)$intergenicRatio %>% as.numeric
```

```{r process_starsolo, include = FALSE}
STARSolo <- Read10X(data.dir=file.path(params$starsolo_matrixDir))
seuratData <- CreateSeuratObject(counts=STARSolo,
                             min.cells=0,
                             min.features=0, project=params$sampleName)
seuratData[['run']]=params$sampleName
seuratData[["percent.mt"]]=PercentageFeatureSet(object=seuratData, pattern = "MT-|mt-")

## Plot the genes counts, RNAs counts and mitochondrial proportions.
qcPlot <- VlnPlot(object=seuratData, features=c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol=3, group.by = "run")

## Filter by gene counts and mitochondrial percentage
## Check the number of cells left first, or subset will throw an error
filtered_cells <- seuratData@meta.data %>% 
    filter(nFeature_RNA > max(200, as.numeric(params$nFeature_RNA_lower)),
           percent.mt < as.numeric(params$percent_mt_cutoff)) %>% 
           nrow
message("filtered_cells is ", filtered_cells)
if(filtered_cells>0){
    seuratData <- subset(
        seuratData, 
        nFeature_RNA > as.numeric(params$nFeature_RNA_lower) & percent.mt < as.numeric(params$percent_mt_cutoff)
    )
    filtered_cells <- seuratData@meta.data %>% nrow
}

## Only perform clustering when more than 150 cells detected
if(filtered_cells >= 150){
    ## Normalization and scaling
    ##seuratData <- NormalizeData(object=seuratData)
    ##seuratData <- FindVariableFeatures(seuratData)
    ##seuratData <- ScaleData(seuratData, vars.to.regress=c("nCount_RNA", "percent.mt"), features=rownames(seuratData))
    
    ## replace above commands with sctransform
    ## default use top 3000 variable features
    seuratData <- SCTransform(seuratData, vars.to.regress = "percent.mt", conserve.memory = TRUE)
    
    seuratData <- RunPCA(object=seuratData, npcs=100)
    
    ## set pca_dim to be used, 30
    pca_dim <- as.numeric(params$pca_dim)
    ## set resolution
    resolution <- as.numeric(params$resolution)
    
    seuratData <- FindNeighbors(object=seuratData, reduction="pca", dims=1:pca_dim)
    seuratData <- FindClusters(object=seuratData, resolution=resolution)
    seuratData <- RunUMAP(object=seuratData, reduction="pca", dims=1:pca_dim)
    
    ##Idents(object=seuratData) <- paste("RNA_snn_res.", resolution, sep="")
    ##clusterPlot <- DimPlot(object=seuratData, reduction="umap", label=TRUE, repel=TRUE)
}
```

```{css echo = FALSE}
.chart-title {
  font-size: 1.2rem;
  font-weight: 700;
  margin-left: 10px;
  margin-right: 10px;
}
```

Mapping Result {data-orientation=rows}
====================

<h2 style="margin-top: 0.5rem; font-weight: bold;">Sample: `r params$sampleName`</h2>

Row
----------

### Estimated Number of Cells

```{r}
cellNum <- starsolo_summary %>% 
    filter(item == "Estimated Number of Cells") %>% 
    pull(value)
valueBox(cellNum, icon = "fa-circle-notch", color = "info")
```

### Mean Reads per Cell

```{r}
## Use 10X style calculation instead of starsolo style
totalReads <- starsolo_summary %>% 
    filter(item == "Number of Reads") %>% 
    pull(value)
readsMean <- (as.numeric(str_remove_all(totalReads, ",")) / as.numeric(str_remove_all(cellNum, ","))) %>% 
    round() %>%
    label_comma()()
valueBox(readsMean, icon = "fa-chart-bar", color = "primary")
```

### Median Genes per Cell

```{r}
medianGenePerCell <- starsolo_summary %>% 
    filter(str_detect(item, regex("Median Gene(Full)* per Cell"))) %>% 
    pull(value)
valueBox(medianGenePerCell, icon = "fa-barcode", color = "warning")
```

Row
----------

### Cell Stats `r infoIcon("cellStats")`

```{r}
d <- starsolo_summary %>% 
    filter(item %in% c("Estimated Number of Cells",
                       "Unique Reads in Cells Mapped to Gene",
                       "Unique Reads in Cells Mapped to GeneFull",
                       "Reads in Cells Mapped to Gene",
                       "Fraction of Unique Reads in Cells",
                       "Fraction of Reads in Cells",
                       "Mean Reads per Cell",
                       "Median Reads per Cell",
                       "UMIs in Cells",
                       "Mean UMI per Cell",
                       "Median UMI per Cell",
                       "Mean Gene per Cell",
                       "Mean GeneFull per Cell",
                       "Median Gene per Cell",
                       "Median GeneFull per Cell",
                       "Total Gene Detected",
                       "Total GeneFull Detected"))
d <- d %>% 
    mutate(item = str_replace(item, "Mean Reads per Cell", "Mean Reads per Cell (STARsolo style)"))
idx <- which(d$item == "Mean Reads per Cell (STARsolo style)")
d <- d %>%
    add_row(item = "Mean Reads per Cell (cellRanger style)", value = readsMean, .before = idx)

names(d) <- c("","")
d %>% kbl() %>% kable_paper("hover", html_font = html_font)
```

### Knee Plot `r infoIcon("kneePlot")`

```{r}
umi_dist <- read_tsv(params$starsolo_bc, col_names =c("UMI"))
cellNum <- str_replace(cellNum, ",", "") %>% as.numeric()
umi_dist <-  umi_dist %>%
    mutate(
        cell = c(1:nrow(umi_dist)),
        group = case_when(
            cell <= cellNum ~ "cell",
            TRUE ~ "background"
            )
        )

fig <- plot_ly(data = umi_dist, x = ~cell, y = ~UMI, type = 'scatter', mode = 'lines',
               color = ~group, colors = c('grey', '#002366'),
               hovertemplate = 'UMI: %{y}<extra>%{fullData.name}</extra>',
               line = list(width = 3),
               height = 400, width = 360)
fig <- layout(fig, xaxis = list(type = "log"),yaxis = list(type = "log"), showlegend = FALSE)
config_plotly_fig(fig) %>% toWebGL()

##ggplot(umi_dist, aes(x=cell, y=UMI, color = group)) + 
##    geom_point(alpha = 0.8) +
##    geom_vline(xintercept = cellNum, 
##               linetype = "dashed", color = "grey") +
##    scale_x_log10(n.breaks = 8, labels = scales::label_number_si()) + 
##    scale_y_log10(n.breaks = 5, labels = scales::label_number_si()) +
##    scale_color_manual(values =c("grey", "royalblue")) +
##    xlab("Cell") + ylab("UMI") +
##    theme_bw() + 
##    theme(panel.grid = element_blank(), 
##          axis.text = element_text(color="black"),
##          legend.position = "none")
```

Row
----------

### Sequencing Stats `r infoIcon("sequencingStats")`

```{r}
d <- starsolo_summary %>% 
    filter(item %in% c("Number of Reads",
                       "Reads With Valid Barcodes",
                       "Sequencing Saturation",
                       "Q30 Bases in CB+UMI",
                       "Q30 Bases in RNA read",
                       "Total Gene Detected",
                       "Total GeneFull Detected"))
names(d) <- c("","")
d %>% kbl() %>% kable_paper("hover", html_font = html_font)
                                        
```

### Mapping Stats `r infoIcon("mappingStats")`

```{r}
d <- starsolo_summary %>% 
    filter(item %in% c("Reads Mapped to Genome: Unique+Multiple",
                       "Reads Mapped to Genome: Unique",
                       "Reads Mapped to Gene: Unique+Multiple Gene",
                       "Reads Mapped to GeneFull: Unique+Multiple GeneFull",
                       "Reads Mapped to Gene: Unique Gene",
                       "Reads Mapped to GeneFull: Unique GeneFull")) %>%
    add_row(item = "Reads Mapped to Exonic Regions", value = label_percent(0.01)(exon_ratio)) %>%
    add_row(item = "Reads Mapped to Intronic Regions", value = label_percent(0.01)(intron_ratio)) %>%
    add_row(item = "Reads Mapped to Intergenic Regions", value = label_percent(0.01)(intergenic_ratio))
names(d) <- c("","")
d %>% kbl() %>% kable_paper("hover", html_font = html_font)
```

Cell Cluster {data-orientation=rows}
====================

Row
----------

### Cell Cluster (colored by cell group) `r infoIcon("clusterCellGroups")`

```{r}
## Only plot cluster when there are more than 150 filtered cells
if(filtered_cells >= 150){
    ##clusterPlot <- DimPlot(object=seuratData, reduction="umap", label=TRUE, repel=TRUE)
    ## extract data from seuratData
    umap_data <- Embeddings(object = seuratData[["umap"]])[colnames(seuratData), c(1, 2)]
    clusterData <- merge(umap_data, seuratData@meta.data, by=0) %>% 
        dplyr::rename("bc" = `Row.names`)
    cellCount <- clusterData %>% group_by(seurat_clusters) %>% summarise(cellCount = n())
    totalCell <- cellCount %>% pull(cellCount) %>% sum()
    clusterData <- clusterData %>% left_join(cellCount, by = "seurat_clusters") %>%
        mutate(totalCell=totalCell,
               text = paste0(cellCount, "/", totalCell))
    
    plot_ly(clusterData, x = ~UMAP_1, y= ~UMAP_2, color = ~seurat_clusters, 
            text = ~text,
            type = 'scatter', mode = 'markers',
            colors = 'Dark2',
            hovertemplate = 'cells: %{text}<extra><b>Cluster: %{fullData.name}</b></extra>') %>%
        config_plotly_fig() %>%
        toWebGL()
    
    ##clusterPlot %>% ggplotly() %>%
    ##    config(displaylogo = FALSE,
    ##           modeBarButtonsToRemove = c('zoom', 'pan', 'select', 'zoomIn', 'zoomOut', 'autoScale',
    ##                                      'hoverClosestCartesian', 'hoverCompareCartesian')) %>% toWebGL()
}else{
    cat("Too few cells: ", filtered_cells, "(filtered by nFeature >= 200)\n")
}
```

### Cell Cluster (colored by read depth) `r infoIcon("clusterUMIDensity")`


```{r}
if(filtered_cells >= 150){
clusterData <- clusterData %>%
    mutate(log10_UMI = log10(nCount_RNA))
    
plot_ly(clusterData, x = ~UMAP_1, y= ~UMAP_2, color = ~log10_UMI,  text = ~nCount_RNA,
        type = 'scatter', mode = 'markers',
        hovertemplate = 'UMI: %{text}<extra></extra>') %>%
    config_plotly_fig() %>%
    toWebGL()

##depthPlot <- FeaturePlot(object = seuratData, features = 'log_nCount_RNA') + scale_color_continuous(name="log10 Read Counts", type = "viridis", direction = 1) + labs(title = NULL)
##depthPlot %>% ggplotly() %>%
##    config(displaylogo = FALSE, 
##           modeBarButtonsToRemove = c('zoom', 'pan', 'select', 'zoomIn', 'zoomOut', 'autoScale',
##                                      'hoverClosestCartesian', 'hoverCompareCartesian')) %>% toWebGL()
}else{
    cat("Too few cells: ", filtered_cells, "(filtered by nFeature >= 200)\n")
}
```

Row {style="margin-right: 8px"}
----------

### Marker Genes of Clusters (top5) `r infoIcon("DEGTable")` {style="width: 100% !important"}

```{r}
if(filtered_cells >= 150){
    markerGenes <- FindAllMarkers(seuratData, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
    if("cluster" %in% colnames(markerGenes)){
        markerGenes %>%
            group_by(cluster) %>%
            slice_max(n = 5, order_by = avg_log2FC) %>% 
            dplyr::relocate(gene, .before=1) %>%
            mutate(p_val = signif(p_val, 3),
                   avg_log2FC = signif(avg_log2FC, 3),
                   p_val_adj = signif(p_val_adj, 3)) %>%
            DT::datatable(rownames = FALSE, width = "100%",
                          options = list(scrollX = T, scrollY = F, pageLength = 10))
    }else{
        cat("No DEGs detected.\n")
    }
}else{
    cat("Too few cells: ", filtered_cells, "(filtered by nFeature >= 200)\n")
}
```

```{r include=FALSE}
if(filtered_cells >= 150 && ("cluster" %in% colnames(markerGenes))){
    markerGenes %>%
        group_by(cluster) %>%
        dplyr::relocate(gene, .before=1) %>%
        write_tsv(paste0(params$sampleName, "_DEG.tsv"))
}
```

Row
----------

### Cell Metrics `r infoIcon("cellMetrics")`

```{r}
qcPlot
```

### Gene Coverage `r infoIcon("geneCoverage")`

```{r}
dep <- read_tsv(params$geneCoverage, skip=3, col_names = NULL) %>% apply(2, sum, na.rm = TRUE)
geneCoverage_df <- tibble(pos = 1:length(dep), dep = dep)
ggplot(geneCoverage_df, aes(x=`pos`, y=`dep`)) +
    geom_line(size=1, color = "#551a8b", alpha=0.8) +
    scale_y_continuous(labels = label_number(scale_cut = cut_short_scale())) +
    xlab("Transcript Relative Position")+ylab("Read Coverage")+
    theme_bw() + theme(panel.grid = element_blank())
```


Row
----------

### Sequencing Saturation `r infoIcon("sequencingDepthCurve")`

```{r}
saturation_data <- fromJSON(params$saturation_json)
saturation_data <- saturation_data[["saturation_data"]] %>%
    as_tibble() %>%
    mutate(across(.col = everything(), .fns = as.numeric)) %>%
    arrange(percentage)

plot_ly(saturation_data, x = ~reads, y= ~saturation,
        type = 'scatter', mode = 'lines',
        hovertemplate = 'Read depth: %{x}<br>Saturation: %{y}<extra></extra>') %>%
    config_plotly_fig() %>%
    toWebGL()
```

### Median Genes per Cell `r infoIcon("medianGeneCurve")`

```{r}
plot_ly(saturation_data, x = ~meanReadPerCell, y= ~medianGenePerCell,
        type = 'scatter', mode = 'lines',
        hovertemplate = 'Mean read per cell: %{x}<br>Median gene per cell: %{y}<extra></extra>') %>%
    config_plotly_fig() %>%
    toWebGL()
```

```{r include=FALSE}
starsolo_summary %>% 
    filter(item %in% c("Estimated Number of Cells",
                       "Unique Reads in Cells Mapped to Gene",
                       "Unique Reads in Cells Mapped to GeneFull",
                       "Fraction of Unique Reads in Cells",
                       "Mean Reads per Cell",
                       "Median Reads per Cell",
                       "UMIs in Cells",
                       "Mean UMI per Cell",
                       "Median UMI per Cell",
                       "Mean Gene per Cell",
                       "Mean GeneFull per Cell",
                       "Median Gene per Cell",
                       "Median GeneFull per Cell",
                       "Total Gene Detected",
                       "Total GeneFull Detected",
                       "Number of Reads",
                       "Reads With Valid Barcodes",
                       "Sequencing Saturation",
                       "Q30 Bases in CB+UMI",
                       "Q30 Bases in RNA read",
                       "Reads Mapped to Genome: Unique+Multiple",
                       "Reads Mapped to Genome: Unique",
                       "Reads Mapped to Gene: Unique+Multiple Gene",
                       "Reads Mapped to GeneFull: Unique+Multiple GeneFull",
                       "Reads Mapped to Gene: Unique Gene",
                       "Reads Mapped to GeneFull: Unique GeneFull")) %>%
    add_row(item = "Reads Mapped to Exonic Regions", value = label_percent(0.01)(exon_ratio)) %>%
    add_row(item = "Reads Mapped to Intronic Regions", value = label_percent(0.01)(intron_ratio)) %>%
    add_row(item = "Reads Mapped to Intergenic Regions", value = label_percent(0.01)(intergenic_ratio)) %>%
    mutate(sampleID=params$sampleName) %>% 
    pivot_wider(sampleID, names_from = item, values_from= value) %>% 
    write_json(paste0(params$sampleName, ".metrics.json"), pretty=TRUE)
```

Running Info {data-orientation=rows}
====================

Row
----------

### Running Information `r infoIcon("runningInfo")`

```{r}
version_info <- fromJSON(params$version_json)
version_tbl <- tribble(
  ~term, ~value,
  "Sample ID", params$sampleName,
  "Pipeline Version", version_info$pipeline_version,
  "Reference Dir", version_info$referenceDir,
  "Reference GTF", version_info$referenceGTF,
  "STAR Version", version_info$STAR_version,
  "Whitelist Matching", version_info$soloCBmatchWLtype,
  "UMIfiltering", version_info$soloUMIfiltering,
  "UMIdedup", version_info$soloUMIdedup,
  "Cell Calling", version_info$soloCellFilter,
  "includeIntron", version_info$includeIntron,
  "includeMultiReads", version_info$includeMultiReads
)
names(version_tbl) <- c("", "")
version_tbl %>% 
    kbl() %>% 
    kable_paper("hover", html_font = "arial, helvetica, sans-serif")
```

```{js}
// set knee plot margin to auto
$( document ).ready(function() {
    document.querySelector(".plot-container").querySelector(".svg-container").style.margin = "auto";
});

// set table left and right margin
$( document ).ready(function() {
    $( "div.chart-stage:has(table.lightable-paper)" ).each(function(){
        this.style.marginLeft = "10px"; 
        this.style.marginRight = "10px";
    });
});
```

```{r include = FALSE}
## save all summary outputs to a tsv file
starsolo_summary %>% 
    filter(item %in% c("Estimated Number of Cells",
                       "Unique Reads in Cells Mapped to Gene",
                       "Unique Reads in Cells Mapped to GeneFull",
                       "Fraction of Unique Reads in Cells",
                       "Mean Reads per Cell",
                       "Median Reads per Cell",
                       "UMIs in Cells",
                       "Mean UMI per Cell",
                       "Median UMI per Cell",
                       "Mean Gene per Cell",
                       "Mean GeneFull per Cell",
                       "Median Gene per Cell",
                       "Median GeneFull per Cell",
                       "Total Gene Detected",
                       "Total GeneFull Detected",
                       "Number of Reads",
                       "Reads With Valid Barcodes",
                       "Sequencing Saturation",
                       "Q30 Bases in CB+UMI",
                       "Q30 Bases in RNA read",
                       "Reads Mapped to Genome: Unique+Multiple",
                       "Reads Mapped to Genome: Unique",
                       "Reads Mapped to Gene: Unique+Multiple Gene",
                       "Reads Mapped to GeneFull: Unique+Multiple GeneFull",
                       "Reads Mapped to Gene: Unique Gene",
                       "Reads Mapped to GeneFull: Unique GeneFull")) %>%
    add_row(item = "Reads Mapped to Exonic Regions", value = label_percent(0.01)(exon_ratio)) %>%
    add_row(item = "Reads Mapped to Intronic Regions", value = label_percent(0.01)(intron_ratio)) %>%
    add_row(item = "Reads Mapped to Intergenic Regions", value = label_percent(0.01)(intergenic_ratio)) %>%
    add_row(item = "SampleID", value = params$sampleName, .before = 1) %>%
    write_tsv(paste0(params$sampleName, ".metrics.tsv"))
```
